File: ui/web-app/src/domain/imageViewport.ts
export type ImageViewportState = {
  left: number;
  top: number;
  zoom: number;
};

export function clampImageZoom(zoom: number) {
  return Math.max(1, Math.min(zoom, 8));
}

export function createInitialImageViewport(
  imageWidth: number,
  imageHeight: number,
  viewportWidth: number,
  viewportHeight: number,
): ImageViewportState {
  const fitScale = Math.min(viewportWidth / imageWidth, viewportHeight / imageHeight);
  const width = imageWidth * fitScale;
  const height = imageHeight * fitScale;
  return {
    left: (viewportWidth - width) / 2,
    top: (viewportHeight - height) / 2,
    zoom: 1,
  };
}

export function imageDisplaySize(
  imageWidth: number,
  imageHeight: number,
  viewportWidth: number,
  viewportHeight: number,
  zoom: number,
) {
  const fitScale = Math.min(viewportWidth / imageWidth, viewportHeight / imageHeight);
  const scale = fitScale * zoom;
  return {
    width: imageWidth * scale,
    height: imageHeight * scale,
  };
}

export function clampImageViewport(
  state: ImageViewportState,
  imageWidth: number,
  imageHeight: number,
  viewportWidth: number,
  viewportHeight: number,
  overscrollPx: number,
): ImageViewportState {
  const size = imageDisplaySize(imageWidth, imageHeight, viewportWidth, viewportHeight, state.zoom);
  const minLeft = viewportWidth - overscrollPx - size.width;
  const maxLeft = overscrollPx;
  const minTop = viewportHeight - overscrollPx - size.height;
  const maxTop = overscrollPx;
  return {
    left: clampToRange(state.left, minLeft, maxLeft),
    top: clampToRange(state.top, minTop, maxTop),
    zoom: clampImageZoom(state.zoom),
  };
}

export function dragImageViewport(
  state: ImageViewportState,
  dx: number,
  dy: number,
  imageWidth: number,
  imageHeight: number,
  viewportWidth: number,
  viewportHeight: number,
  overscrollPx: number,
) {
  return clampImageViewport(
    {
      ...state,
      left: state.left + dx,
      top: state.top + dy,
    },
    imageWidth,
    imageHeight,
    viewportWidth,
    viewportHeight,
    overscrollPx,
  );
}

export function zoomImageAroundPoint(
  state: ImageViewportState,
  anchorX: number,
  anchorY: number,
  nextZoom: number,
  imageWidth: number,
  imageHeight: number,
  viewportWidth: number,
  viewportHeight: number,
  overscrollPx: number,
) {
  const zoom = clampImageZoom(nextZoom);
  const scale = zoom / state.zoom;
  return clampImageViewport(
    {
      left: anchorX - (anchorX - state.left) * scale,
      top: anchorY - (anchorY - state.top) * scale,
      zoom,
    },
    imageWidth,
    imageHeight,
    viewportWidth,
    viewportHeight,
    overscrollPx,
  );
}

function clampToRange(value: number, min: number, max: number) {
  if (min > max) {
    return (min + max) / 2;
  }
  return Math.min(max, Math.max(min, value));
}

File: ui/android-app/app/src/main/java/org/aerobag/app/domain/ImageViewport.kt
package org.aerobag.app.domain

data class ImageViewportState(
    val leftPx: Float,
    val topPx: Float,
    val zoom: Float,
)

data class ImageDisplaySize(
    val widthPx: Float,
    val heightPx: Float,
)

fun clampImageZoom(zoom: Float): Float = zoom.coerceIn(1f, 8f)

fun createInitialImageViewport(
    imageWidthPx: Float,
    imageHeightPx: Float,
    viewportWidthPx: Float,
    viewportHeightPx: Float,
): ImageViewportState {
    val fitScale = minOf(viewportWidthPx / imageWidthPx, viewportHeightPx / imageHeightPx)
    val width = imageWidthPx * fitScale
    val height = imageHeightPx * fitScale
    return ImageViewportState(
        leftPx = (viewportWidthPx - width) / 2f,
        topPx = (viewportHeightPx - height) / 2f,
        zoom = 1f,
    )
}

fun imageDisplaySize(
    imageWidthPx: Float,
    imageHeightPx: Float,
    viewportWidthPx: Float,
    viewportHeightPx: Float,
    zoom: Float,
): ImageDisplaySize {
    val fitScale = minOf(viewportWidthPx / imageWidthPx, viewportHeightPx / imageHeightPx)
    val scale = fitScale * zoom
    return ImageDisplaySize(
        widthPx = imageWidthPx * scale,
        heightPx = imageHeightPx * scale,
    )
}

fun clampImageViewport(
    state: ImageViewportState,
    imageWidthPx: Float,
    imageHeightPx: Float,
    viewportWidthPx: Float,
    viewportHeightPx: Float,
    overscrollPx: Float,
): ImageViewportState {
    val size = imageDisplaySize(
        imageWidthPx = imageWidthPx,
        imageHeightPx = imageHeightPx,
        viewportWidthPx = viewportWidthPx,
        viewportHeightPx = viewportHeightPx,
        zoom = state.zoom,
    )
    val minLeft = viewportWidthPx - overscrollPx - size.widthPx
    val maxLeft = overscrollPx
    val minTop = viewportHeightPx - overscrollPx - size.heightPx
    val maxTop = overscrollPx
    return ImageViewportState(
        leftPx = clampToRange(state.leftPx, minLeft, maxLeft),
        topPx = clampToRange(state.topPx, minTop, maxTop),
        zoom = clampImageZoom(state.zoom),
    )
}

fun dragImageViewport(
    state: ImageViewportState,
    dxPx: Float,
    dyPx: Float,
    imageWidthPx: Float,
    imageHeightPx: Float,
    viewportWidthPx: Float,
    viewportHeightPx: Float,
    overscrollPx: Float,
): ImageViewportState = clampImageViewport(
    state = ImageViewportState(
        leftPx = state.leftPx + dxPx,
        topPx = state.topPx + dyPx,
        zoom = state.zoom,
    ),
    imageWidthPx = imageWidthPx,
    imageHeightPx = imageHeightPx,
    viewportWidthPx = viewportWidthPx,
    viewportHeightPx = viewportHeightPx,
    overscrollPx = overscrollPx,
)

fun zoomImageAroundPoint(
    state: ImageViewportState,
    anchorX: Float,
    anchorY: Float,
    nextZoom: Float,
    imageWidthPx: Float,
    imageHeightPx: Float,
    viewportWidthPx: Float,
    viewportHeightPx: Float,
    overscrollPx: Float,
): ImageViewportState {
    val zoom = clampImageZoom(nextZoom)
    val scale = zoom / state.zoom
    return clampImageViewport(
        state = ImageViewportState(
            leftPx = anchorX - (anchorX - state.leftPx) * scale,
            topPx = anchorY - (anchorY - state.topPx) * scale,
            zoom = zoom,
        ),
        imageWidthPx = imageWidthPx,
        imageHeightPx = imageHeightPx,
        viewportWidthPx = viewportWidthPx,
        viewportHeightPx = viewportHeightPx,
        overscrollPx = overscrollPx,
    )
}

private fun clampToRange(value: Float, min: Float, max: Float): Float {
    val lower = minOf(min, max)
    val upper = maxOf(min, max)
    return value.coerceIn(lower, upper)
}
